common/quotas/global/algorithm/requestweighted_test.go

```diff
@@ -0,0 +1,746 @@
+		h1: 10 / 1209.0,  // 0.07698229407
+		h2: 200 / 1209.0, // 0.1539645881
+		h3: 999 / 1209.0, // 0.7690531178
+	}, floaty(weights), 0.1, "should be close to true load balance")
+}
+
+// converts for testify/assert as it requires same types, not just same underlying type
+func floaty[K comparable, V numeric](m map[K]V) map[K]float64 {
+	out := make(map[K]float64, len(m))
+	for k, v := range m {
+		out[k] = float64(v)
+	}
+	return out
+}
+
+func TestConcurrent(t *testing.T) {
+	// essentially a fuzz-test for race purposes.
+	// values aren't checked, but it shouldn't panic / shouldn't race / etc.
+	// low timeout + real time clock to also have wall-clock changes race with logic.
+	//
+	// this test occasionally covers every line all on its own (currently),
+	// though it's not guaranteed.
+
+	// should be "much" longer than expected update rate
+	const (
+		updateRate      = time.Millisecond // fairly arbitrary, max gap between updates
+		targetDuration  = 100 * updateRate // also minimum number of updates
+		numHosts        = 10
+		numUpdaters     = 10           // fairly arbitrary
+		updatesPerBatch = numHosts / 3 // intentionally below len(hosts) to allow some to gc normally
+	)
+
+	agg, _ := newForTest(t, configSnapshot{
+		weight:    0.5,
+		rate:      updateRate,
+		maxMissed: 2,
+		gcAfter:   3, // relatively low to trigger implicit gc, check coverage if changing the values
+	})
+	agg.clock = clock.NewRealTimeSource()
+
+	var hosts []Identity
+	for i := 0; i < numHosts; i++ {
+		hosts = append(hosts, Identity(fmt.Sprintf("host %d", i)))
+	}
+	var keys []Limit
+	for i := 'a'; i <= 'z'; i++ {
+		keys = append(keys, Limit(i))
+	}
+
+	do := make(chan struct{}, numUpdaters)
+	var done sync.WaitGroup
+
+	// run some goroutines to update/read in batches
+	for i := 0; i < numUpdaters; i++ {
+		done.Add(1)
+		go func() {
+			defer done.Done()
+
+			for {
+				_, ok := <-do
+				if !ok {
+					// chan's closed, stop
+					return
+				}
+				num := rand.Intn(len(keys))
+				updates := make(map[Limit]Requests, num)
+				for i := 0; i < num; i++ {
+					updates[keys[rand.Intn(len(keys))]] = Requests{
+						Accepted: rand.Intn(100),
+						Rejected: rand.Intn(100),
+					}
+				}
+				host := hosts[rand.Intn(len(hosts))]
+
+				// randomly update or read
+				if rand.Intn(2) == 0 {
+					agg.Update(host, updates, updateRate)
+				} else {
+					agg.HostWeights(host, maps.Keys(updates))
+				}
+			}
+		}()
+	}
+
+	// run a "trigger some work occasionally" goroutine
+	done.Add(1)
+	start := time.Now()
+	go func() {
+		defer done.Done()
+
+		for {
+			if time.Since(start) > targetDuration {
+				close(do)
+				return
+			}
+			// sleep a random portion of the update rate
+			time.Sleep(time.Duration(rand.Intn(int(updateRate))))
+			// allow a random number of updates.
+			// should be non-blocking to further encourage racing, when possible
+			for i := rand.Intn(updatesPerBatch); i > 0; i-- {
+				do <- struct{}{}
+			}
+		}
+	}()
+
+	done.Wait()
```

Comment:
ah, TIL we have a thing for that.  yea, makes sense here since I don't actually need anything it's protecting, it's just waiting for the [hammer it concurrently] to complete.

I do hope we don't use it outside of tests though :\  using things like that often implies "intentionally allow race conditions"...